File: src/common/f1tenth_interfaces/test/test_actuator_command.cpp
// Copyright 2026 F1TENTH Racing Stack contributors
//
#include <gtest/gtest.h>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/serialization.hpp>

#include "f1tenth_interfaces/msg/actuator_command.hpp"

TEST(ActuatorCommandSerialization, round_trip_preserves_fields) {
  f1tenth_interfaces::msg::ActuatorCommand msg;
  msg.header.stamp.sec = 42;
  msg.header.stamp.nanosec = 123456789;
  msg.header.frame_id = "base_link";
  msg.generation = 7;
  msg.observation_stamp.sec = 41;
  msg.observation_stamp.nanosec = 987654321;
  msg.drive_current_a = 3.5;
  msg.brake_current_a = 0.0;
  msg.servo_position = 0.4495;
  msg.longitudinal = 0.25;
  msg.steering = -0.12;
  msg.source = f1tenth_interfaces::msg::ActuatorCommand::SOURCE_RL;

  rclcpp::Serialization<f1tenth_interfaces::msg::ActuatorCommand> serializer;
  rclcpp::SerializedMessage serialized;
  serializer.serialize_message(&msg, &serialized);

  f1tenth_interfaces::msg::ActuatorCommand restored;
  serializer.deserialize_message(&serialized, &restored);

  EXPECT_EQ(restored.header.stamp.sec, 42);
  EXPECT_EQ(restored.header.stamp.nanosec, 123456789u);
  EXPECT_EQ(restored.header.frame_id, "base_link");
  EXPECT_EQ(restored.generation, 7u);
  EXPECT_EQ(restored.observation_stamp.sec, 41);
  EXPECT_EQ(restored.observation_stamp.nanosec, 987654321u);
  EXPECT_DOUBLE_EQ(restored.drive_current_a, 3.5);
  EXPECT_DOUBLE_EQ(restored.brake_current_a, 0.0);
  EXPECT_DOUBLE_EQ(restored.servo_position, 0.4495);
  EXPECT_FLOAT_EQ(restored.longitudinal, 0.25f);
  EXPECT_FLOAT_EQ(restored.steering, -0.12f);
  EXPECT_EQ(restored.source, f1tenth_interfaces::msg::ActuatorCommand::SOURCE_RL);
}

TEST(ActuatorCommandConstants, source_values_are_stable) {
  EXPECT_EQ(f1tenth_interfaces::msg::ActuatorCommand::SOURCE_SAFE, 0);
  EXPECT_EQ(f1tenth_interfaces::msg::ActuatorCommand::SOURCE_RL, 1);
  EXPECT_EQ(f1tenth_interfaces::msg::ActuatorCommand::SOURCE_TELEOP, 2);
  EXPECT_EQ(f1tenth_interfaces::msg::ActuatorCommand::SOURCE_SAFETY, 3);
}

File: src/common/f1tenth_interfaces/test/test_actuator_command.py
"""Serialization and source-enum parity for ActuatorCommand."""

import pytest
from builtin_interfaces.msg import Time
from f1tenth_interfaces.msg import ActuatorCommand
from rclpy.serialization import deserialize_message, serialize_message
from std_msgs.msg import Header


def test_actuator_command_serialization_round_trip():
    msg = ActuatorCommand()
    msg.header = Header(stamp=Time(sec=10, nanosec=500), frame_id="base_link")
    msg.generation = 99
    msg.observation_stamp = Time(sec=9, nanosec=250)
    msg.drive_current_a = 0.0
    msg.brake_current_a = 5.0
    msg.servo_position = 0.12
    msg.longitudinal = -0.5
    msg.steering = 0.33
    msg.source = ActuatorCommand.SOURCE_SAFETY

    restored = deserialize_message(serialize_message(msg), ActuatorCommand)

    assert restored.header.stamp.sec == 10
    assert restored.header.stamp.nanosec == 500
    assert restored.header.frame_id == "base_link"
    assert restored.generation == 99
    assert restored.observation_stamp.sec == 9
    assert restored.observation_stamp.nanosec == 250
    assert restored.drive_current_a == 0.0
    assert restored.brake_current_a == 5.0
    assert restored.servo_position == pytest.approx(0.12)
    assert restored.longitudinal == pytest.approx(-0.5)
    assert restored.steering == pytest.approx(0.33)
    assert restored.source == ActuatorCommand.SOURCE_SAFETY


def test_actuator_command_source_constants():
    assert ActuatorCommand.SOURCE_SAFE == 0
    assert ActuatorCommand.SOURCE_RL == 1
    assert ActuatorCommand.SOURCE_TELEOP == 2
    assert ActuatorCommand.SOURCE_SAFETY == 3
